dvc/pkg/__init__.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import unicode_literals
+
+import os
+import copy
+import shutil
+
+from dvc.utils.compat import urlparse, str, builtin_str
+
+from dvc.config import Config
+from dvc.remote.config import remote_add
+
+from .config import get_pkg_settings
+
+
+class Pkg(object):
+    def __init__(self, repo, settings):
+        self.settings = copy.copy(settings)
+
+        name = None
+        if isinstance(settings, str) or isinstance(settings, builtin_str):
+            name = settings
+            settings = {}
+        elif isinstance(settings, dict):
+            name = settings.pop("name")
+
+        if os.path.exists(name) or urlparse(name).scheme:
+            if Config.SECTION_PKG_URL not in settings:
+                settings[Config.SECTION_PKG_URL] = name
+            name = os.path.basename(name)
+            config_settings = {}
+        else:
+            config_settings = (
+                get_pkg_settings(repo.config, name) if name else {}
+            )
+
+        config_settings.update(settings)
+
+        self.url = config_settings.get(Config.SECTION_PKG_URL)
+        self.revision = config_settings.get(Config.SECTION_PKG_REVISION)
+
+        self._dir = os.path.join(repo.pkg.pkg_dir, name)
+
+        self._repo = None
+
+    @property
+    def repo(self):
+        if not self._repo or self._repo.root_dir != self._dir:
+            from dvc.repo import Repo
+
+            revision = self.revision
+            if revision:
+                revision = "remotes/origin/" + revision
+
+            self._repo = Repo(self._dir, revision)
+
+        return self._repo
+
+    @property
+    def installed(self):
+        return os.path.isdir(self._dir)
+
+    def _setup_remote(self):
+        """Setup remote pointing to the original repository if default remote
+        is not setup already. For example, say you have a local dvc project
+        at `/home/user/dvc/myproj` and now you want to use it as a pkg in your
+        another project. If `myproj` doesn't have default remote setup, then
+        we can simply use it's cache as a local remote to pull from.
+        """
+        from dvc.repo import Repo
+
+        core = self.repo.config.config.get(Config.SECTION_CORE, {})
+        remote_name = core.get(Config.SECTION_CORE_REMOTE)
+
+        if remote_name or not os.path.isdir(self.url):
+            return
+
+        pkg_repo = Repo(self.url)
+        remote_url = os.path.basename(self.url)
+
+        # NOTE: setting default remote in local config, so it works for
+        # any branch, since local config is not included into git repo
+        remote_add(
```

Comment:
this looks overcomplicated. I think package should be cloned into `dvc/pkg` and its cache should be independent from the source. Either we can reuse the destination's cache, or create a new one per package. If destination package does not define a remote it's too advanced case to focus initially, and probably should not be solved by modifying config.

Reply:
This is convenient when you want to use your local project as a package in another project. This was one of the basic scenarios that we were talking about.

Reply:
I'm actually not sure if this is a basic scenario. The basic one is to attach something from the github repo. @dmpetrov what are your thoughts on this?

Anyway, it looks like a hack. It's hard to explain, there is a potential clash, etc.

If there is an intention of reusing the cache between projects I would just recommend users define a shared cache for them.

Reply:
@shcheklein right, the basic scenario we should be targeting is an external repo.

Reply:
I still think that it is useful and not confusing. I could totally see someone asking for "could I use data from my local project without me having to setup an external remote?" External repos are supported anyways.

Reply:
I agree, but at what cost? My main problem with that we are introducing some hacks to do this and it looks like this scenario drives to some extent implementation decisions for the main one - external repos. While it should be other way around - external first (you will have to clone, reuse cache, most likely there will no be problems with dirs and other stuff that look like comes from this optimization), then local, then think about certain optimizations.

Reply:
it's confusing - it will be very hard to explain what's going on - even single repo cache/remote/workspace relationship is super advanced and very hard explainable stuff. Setting up a remote to an external cache? This is super advanced. Feels like a hack (side effects to config - potential conflicts, etc).

Reply:
I don't see where this is hurting external repos. If no default remote is detected -- then we use the cache dir of the package itself as a remote. This is only done for local repos that don't have default remotes setup. Pretty specific case and in most situations it is going to be what users expect when tryin to import something from their local repo.

Reply:
so, what it's not even local repo - it's a local repo with a cache only? do we need to support it all? I just don't see any reasons introducing this hack now. You want to share - use shared cache or identify remote.

Reply:
I think we do need to support it. For example
```
cd ~
mkdir datasets
cd datasets
git init
dvc init
echo foo > foo
dvc add foo

cd ~
mkdir experiment
cd experiment
git init
dvc init
dvc pkg import ~/datasets foo
```

No remote required. As you've said, remotes are advanced stuff, and the trick that you are referring to is actually hiding it to just do what user expects from dvc.

Reply:
After IRL discussion I agree that this is needlessly complicated "feature". Need to use external cache in that scenario and try to not copy cache when fetching by using main projects cache,